internal/provider/folder_data_source.go

```diff
@@ -0,0 +1,181 @@
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/hashicorp/terraform-plugin-framework/datasource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+// Ensure the implementation satisfies the expected interfaces.
+var (
+	_ datasource.DataSource              = &folderDataSource{}
+	_ datasource.DataSourceWithConfigure = &folderDataSource{}
+)
+
+type folderDataSource struct {
+	provider *provider
+}
+
+func (d *folderDataSource) Schema(
+	_ context.Context, _ datasource.SchemaRequest, resp *datasource.SchemaResponse,
+) {
+	resp.Schema = schema.Schema{
+		Description: "A CockroachDB Cloud folder. Folders can contain clusters or other folders.  They can be used to group resources together for the purposes of access control, organization or fine grained invoicing.",
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				MarkdownDescription: "The id the folder.",
+				Optional:            true,
+			},
+			"path": schema.StringAttribute{
+				MarkdownDescription: "An absolute path to the folder. Trailing slashes are optional. (i.e. /folder1/folder2)",
+				Optional:            true,
+			},
+			"name": schema.StringAttribute{
+				MarkdownDescription: "Name of the cluster.",
+				Computed:            true,
+			},
+			"parent_id": schema.StringAttribute{
+				Computed:            true,
+				MarkdownDescription: "The ID of the folders's parent folder. 'root' is used for a folder at the root level.",
+			},
+		},
+	}
+}
+
+func (d *folderDataSource) Metadata(
+	_ context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse,
+) {
+	resp.TypeName = req.ProviderTypeName + "_folder"
+}
+
+func (d *folderDataSource) Configure(
+	_ context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse,
+) {
+	if req.ProviderData == nil {
+		return
+	}
+	var ok bool
+	if d.provider, ok = req.ProviderData.(*provider); !ok {
+		resp.Diagnostics.AddError("Internal provider error",
+			fmt.Sprintf("Error in Configure: expected %T but got %T", provider{}, req.ProviderData))
+	}
+}
+
+func (d *folderDataSource) Read(
+	ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse,
+) {
+	if d.provider == nil || !d.provider.configured {
+		addConfigureProviderErr(&resp.Diagnostics)
+		return
+	}
+
+	var folderDataSource FolderDataSourceModel
+	diags := req.Config.Get(ctx, &folderDataSource)
+
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		resp.Diagnostics.AddWarning("Error loading the folder", "")
+		return
+	}
+
+	if folderDataSource.ID.IsNull() && folderDataSource.Path.IsNull() {
+		resp.Diagnostics.AddError(
+			"Unable to lookup folder",
+			"At least one of ID or Path must be set, both were empty",
+		)
+		return
+	}
+
+	var folder *client.FolderResource
+	if !folderDataSource.ID.IsNull() {
+		folderID := folderDataSource.ID.ValueString()
+
+		if !uuidRegex.MatchString(folderID) {
```

Comment:
Interesting that we have this. Maybe we should choose between using the regex and using `uuid.Parse`

Reply:
hmm, yeah.  the mix of these is no bueno.  I'll send a followup PR to replace the regexp with `uuid.Parse`